### DISCUSSION TITLE
What to choose as a new framework for Pd GUI?

### DISCUSSION THREAD

**Discussion by Ant1r** [pure-data/pure-data#1844] [Ideas]
I would like to start the discussion about the choice of a new library for the Pd GUI.

First, it could be useful to know what are the motivations for changing the Pd GUI to another GUI toolkit/framework/language?
Here are mine:
1. faster interface even with large patches
2. better look (anti-aliasing) - better rendering consistency between OSes (e.g Tk is antialiased on Macos, not on Linux).
3. zoom / scroll (better adaptation to display dpi differences, and to more than 50 years old eyes...).
4. possibility for more GUIs (I'm thinking mainly of GUI externals here, as it's good to keep a small core): image, oscilloscope, wave table viewer with zoom / scroll / selection, spectrum analyzer, timeline, sequencer...
5. allow dynamic creation of GUIs by users, why not with the help of a minimal scripting language (e.g Lua)
6. possibility for displaying on mobiles (and web): that would save ports to mobile platforms having to re-implement the GUIs to the native API...

8. light, simple, as few dependencies as possible.

I think that #6 and #7 disqualify big desktop-oriented toolkits like Qt an GTK, which require many dependencies and aren't likely to run on mobiles.

Openframeworks is also a bit too large for this task I think, and depends on a bunch of dependencies too.

Actually I believe that a minimal interface to opengl-like hardware drivers would be the best.

I took a quick look to "Dear Imgui" (which sounds quite funny in the Pd world btw!), see https://github.com/ocornut/imgui ; it would be some work to port the Pd GUI to such interfaces, but it doesn't sounds impossible (several "node editor" implementations already exist). After that, the port to every platform is given for almost free!

The size of the binaries seems small (around 2MB for (not so) simple examples), the source code and the build system are incredibly minimal, and the drawing is quite efficient.

I tried a mock-up of the Pd interface using imgui-node-editor (https://github.com/thedmd/imgui-node-editor) and even with 1000 objects and 1500 connections, the framerate is still ~20 fps on my more-than-10-years-old machine.
There will be a lot to discuss, and it could be much better solutions, but I can't resist to show you what this looks like:

![imgui-pdgui-1-2](https://user-images.githubusercontent.com/3748246/205345851-140879d8-a843-40ff-b83b-2be61992b820.gif)

Here is the code if you want to try:
[imgui-node-editor.zip](https://github.com/pure-data/pure-data/files/10142989/imgui-node-editor.zip)

**Comment by Ant1r** [comment]
correction: Qt seems to support both Android (https://doc.qt.io/qt-5/android.html) and iOS (https://doc.qt.io/qt-5/ios.html). GTK definitely doesn't.

**Comment by Ant1r** [comment]
One problem with imgui is it doesn't support multi-touch, and doesn't plan to do it soon (https://github.com/ocornut/imgui/issues/165). Multi-touch would be required for mobile apps. Maybe multi-touch could be managed by a custom event handler (or even "locally", for only a few GUI objects, like bang, toggle, sliders etc.).

Another problem is that, due to the nature of the "immediate-mode" paradigm, the whole window is redrawn every frame. While that proves to be very efficient, the cost is a constant CPU load, which is quite low for simpler layouts, but can become higher for more complex ones, until it reaches 100% and the framerate drops. 
This constant load is a handicap for battery-powered devices. It could be lowered using a bit of caching, either at draw-list (vertice list) level or at framebuffer level (like drawing every object in a dedicated framebuffer, which would be refreshed only under certain conditions).

**Comment by sebshader** [comment]
I guess qt looks like the best choice. I'm not sure mobile development should be too much of a criterion though since the layout is so much different anyways (editing a patch on mobile sounds.. kinda painful. Plus the view is so much different in being non-fixed size that maybe there should be a different layout for mobile UI anyways, like some css-based thing).

wxwidgets looked good at first, but it doesn't really have good support for a 'canvas'.
I think it's between gtk and qt.

**Reply by Ant1r** [reply]
Of course I don't intend to edit patches on mobiles. 
But Pd has proven to be quite useful as a fast prototyping tool to build mobile apps (see Pd(droid)Party for example). It could be even much better with a proper GUI!

**Comment by sebshader** [comment]
also, Max/MSP switched to juce which maybe should be considered as well

**Reply by umlaeute** [reply]
urgh. i really hate JUCE. i think the framework shouldn't need re-adjusting for every minor release (e.g. ew need to update our code for switching from JUCE-7.0.2 to JUCE-7.0.3, which doesn't even mention any [breaking change](https://github.com/juce-framework/JUCE/blob/7.0.3/BREAKING-CHANGES.txt) in it's docs).

**Reply by sebshader** [reply]
that's fair. from what I've heard gtk gets similar criticism but not quite that bad.
I guess juce devs are just fully into apple-like current optimal design vs. backwards compatibility. But at least in their case you can at least ship with an old version that builds. So most projects probably just stick with a version for a while.
one thing you can't really say about tcl/tk is that it isn't very backwards compatible
edit: kinda looks like they just forgot to document changes to 7.0.3

**Comment by giuliomoro** [comment]
@Ant1r did you have a look for context at https://github.com/pure-data/pure-data/discussions/1695, https://github.com/pure-data/pure-data/pull/1765 , https://github.com/BelaPlatform/pure-data-web-GUI ?

**Comment by zepadovani** [comment]
I advocate for adopting a Web based GUI, using HTML, CSS, and JavaScript (or TypeScript) in a Desktop (and Mobile?) graphical application. These technologies have proven their versatility across various applications, with JavaScript engines exhibiting remarkable advancements in recent years. The extensive toolsets available for JavaScript and TypeScript offer rich options for GUI development and visualization frameworks, catering to a broad spectrum of UI/UX needs.

The capabilities of JavaScript applications have reached impressive heights... a good example: https://threejs.org/examples/. Frameworks such as React.js, Angular, Vue, and similar counterparts are designed to facilitate the creation of reusable components. This approach allows for the efficient construction of UI elements for different objects for PD, using standard GUI components for things like sliders, numbers and non graphical objects, and also allowing to create custom-designed components for more specific visual elements such as arrays and sonograms.

Also, JavaScript is being used to develop different kind of desktop and mobile applications, and even in Desktop Environments (Deepin). The support for new packages and solutions in the JavaScript ecosystem is robust (because it is supported by different organizations and companies), making it an ideal choice for developing applications that require a larger number of developers (I mean, there are not so many people developing things in Tk...). Some frameworks should be considered because they are really wide used, with a strong support of their own developer communities: Electron, NW.js, Cordova, React Native. 

Recently, I am exploring also pywebview, which bridges Python and JS for cross-platform graphical app development: it's really easy to start developing something with a great graphical usability.

Those solutions would probably make it much easier to develop similar tools for PD and other languages/environments (and I do think that this is really important for some external projects!).

Oh, ans also... there are already many patch-based GUI solutions using JS/HTML/CSS: PurrData, obviously, but also things like that: https://idflood.github.io/ThreeNodes.js/index_optimized.html#example/postprocessing1.json. Porbably this would make it a little easier to start developing this new GUI solution...

Well, I know that not everyone likes JS (and don't like very much the verbose syntax), but there are really good reasons to consider using a web-based solution in my point of view.

**Reply by sebshader** [reply]
I disagree. not only are javascript and css abysmal technologies in terms of design, javascript engines are notoriously unpredictable and electron apps and the like contain a ton of bloat that isn't necessary for a gui program.
True, tcl isn't that much better.. but still that's something to get away from.

**Reply by zepadovani** [reply]
Well, Electron is indeed problematic with bloats, but there are other options there. Regarding predictability, JS guis are being used in many situations without major issues (Max, Purrdata...) and, despite the verbosity of the language, it is now much more mature as a tool because it is required everywhere (devops everywhere) and has tons of GUI frameworks, visualization modules, etc. Also, as web pages need to be downloaded and to exhibit things fast, JS/HTML5/CSS GUIs are actually very small in terms of size and load time. 

In any case, I think that the decision about a new GUI should be evidence-based regarding all these considerations (load time, flexibility, stability/predictability, easy integration with different OSes, possibility to build/deliver a patch as a build application, web versions, etc).

**Reply by Spacechild1** [reply]
Coming from classic desktop UI frameworks, web UI dev feels like a Rube Goldberg machine to me... I am really wondering how much of this will be still around in 10 years - or replaced by the next hot thing. For comparison, Qt has been around for over 25 years and thus stood the test of time.

> Well, Electron is indeed problematic with bloats, but there are other options there

Most web UI frameworks still require a browser engine, with React Native being a notable exception.

> JS guis are being used in many situations without major issues (Max, Purrdata...)

Max/MSP uses JUCE (C++).

PurrData uses nwjs (alternative to Electron) and I must say that in my quick tests it performs significantly better than our Tk GUI - at least regarding CPU usage. However, large patches feel comparably sluggish, which I think has more to do with the fact that core and GUI communicate over a TCP connection.

This brings me to an important point: **We should really consider moving the GUI into the main executable**. Historically, running the GUI a separate process was "poor man's multi-threading", but it just creates unnecessary overhead, with literally no upsides.*) A C++ UI framework, such as Qt, can be easily plugged directly into the main Pd application. After all, it's just a bunch of shared libraries. Most web UI frameworks, however, such as Electron or nwjs, *require* a dedicated process, so this wouldn't be possible.

For me at least, the Pd GUI should be simple, lightweight and as fast as possible. This pretty much excludes all web UI frameworks.

---

*) Yes, sometimes you want the GUI to live on a separate machine and communicate over a network connection. This would still be possible, you just have to swap out the underlying communication layer, i.e. replacing the in-process message queue with a TCP socket.

**Reply by zepadovani** [reply]
Oh, I thought Max was also on nw.js (probably I was mislead by the icons I saw in a students project). Thanks for correcting me!

And yes, you have a point regarding Qt. Indeed, it is also very flexible and portable after all.

But... is moving the GUI to the main executable really that important? Is the overhead you mention related to execution or programming tasks? (Sorry, I really want to understand this point better.) 

I ask also because it seems crucial for some projects to have a GUI-free (or agnostic) PD executable. This is also the case, as long as I see, in SuperCollider, where the Qt guis, when used, are run by the language and the audio stuff in the server. In any case, I suppose you are already envisioning something that could be turned off or configured, in special situations, to communicate over the network, right? This brings up the need for a more structured API and perhaps an even better-structured file format for patches.

Regarding JS/web based GUI, the main advantage I see is bringing PD closer to the vast world of GUI implementations already used by artists dealing with interaction. I come from the music/composition/computer music world, but having many students and colleagues from digital and visual arts working with these tools (Processing, p5.js, openFrameworks), it's a natural impulse to bring our tools closer. It's true that JS is still evolving as a technology, but, like what's happening with Python, the tendency, as long as I can see from industry demands, is for it to become more robust and widely used. In any case, I understand the considerations regarding lightweight+simple+fast principles, and I completely agree. I would only add "flexibility", because I think that it would be really important to have an environment that could allow easy development to build different types of GUIs for specific purposes. I think having a language that offers many modules providing this (like vexflow for scores, p5.js for sketches, etc.) would be nice. I mean... creating somehting like a sonogram, an ambisonics visualization GUI or something like that should not be such a pain.

In any case, I will be really happy on the day we get rid of Tk. My feeling is that it's a contradiction to have a visual programming language in which the visual aspect is so limited and limiting.

**Reply by Spacechild1** [reply]
> But... is moving the GUI to the main executable really that important? Is the overhead you mention related to execution or programming tasks? (Sorry, I really want to understand this point better.)

Execution. Every GUI command has to be serialized, transmitted over a TCP connection, deserialized on the other side and then finally executed. Also, Pd intentionally throttles the network I/O. If the GUI was in the same process, you would just push commands to a (thread-safe) queue.

> I ask also because it seems crucial for some projects to have a GUI-free (or agnostic) PD executable. 

You would be able to compile Pd without internal GUI. (You can do the same for sclang, btw.)

> In any case, I suppose you are already envisioning something that could be turned off or configured, in special situations, to communicate over the network, right?

Right, my idea is that the GUI commands could be send to another process via a TCP connection, if desired.

> This brings up the need for a more structured API and perhaps an even better-structured file format for patches.

I don't see how this could possibly affect the patch file format...

> I think having a language that offers many modules providing this (like vexflow for scores, p5.js for sketches, etc.) would be nice.

Visuals are simply not the main focus of Pd. Also, not all of these apps/frameworks are built with JS (Processing, OpenFrameworks, Touch Designer, VVVV, Unity, Unreal, Love3D, etc.) Pd can communicate with any of these programs via OSC.

> because I think that it would be really important to have an environment that could allow easy development to build different types of GUIs for specific purposes.

You *can* make your own custom UIs in Pd. The problem is that currently you need to send raw Tcl commands. If we manage to abstract all the various drawing commands, then this would become much more practical.

> My feeling is that it's a contradiction to have a visual programming language in which the visual aspect is so limited and limiting.

Well put.

**Comment by tremblap** [comment]
For me, as the main FluCoMa@pd maintenance officer (sic) what kills my joy is how I cannot draw a few hundred points without cpu hogging. Obviously, suffering the tk-c api is a pain, but if at least I had some results! So consistency between oses and streamlined efficient object-based drawing is what I'm hoping for... the prototype of @Ant1r made me drool I have to say :)

I'm surprised nobody talks about SuperCollider - they have very similar problems (free, cross-platform, etc) and have embraced Qt to some joy, no?

**Comment by Spacechild1** [comment]
> I'm surprised nobody talks about SuperCollider 

I have mentioned SuperCollider in other related threads in the past. It is indeed a good example for what Qt can deliver.
